tests/test_logging.py

```diff
@@ -0,0 +1,323 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import inspect
+import logging
+import os
+import tempfile
+import unittest
+from dataclasses import dataclass
+from logging import LogRecord
+
+from accelerate import Accelerator
+from accelerate.logging import MultiProcessAdapter, get_logger
+from accelerate.state import PartialState
+from accelerate.test_utils.testing import (
+    DEFAULT_LAUNCH_COMMAND,
+    device_count,
+    execute_subprocess_async,
+    require_multi_device,
+)
+
+
+MESSAGE_SEP = " - message:"
+LOG_FORMAT = "%(levelname)s:%(filename)s:%(name)s:%(lineno)d:%(funcName)s" + MESSAGE_SEP + "%(message)s"
+
+
+# Like `LogRecord`, but not the same thing
+@dataclass
+class DecodedLogRecord:
+    levelname: str
+    filename: str
+    name: str
+    lineno: int
+    funcName: str
+    msg: str  # encoded message
+
+    @classmethod
+    def decode(cls, line: str):
+        assert bool(line), "empty line"
+        prefix, message = line.split(MESSAGE_SEP)
+        levelname, filename, name, lineno, funcName = prefix.split(":")
+        return cls(
+            levelname=levelname,
+            filename=filename,
+            name=name,
+            lineno=int(lineno),
+            funcName=funcName,
+            msg=message,
+        )
+
+
+@dataclass
+class CallSite:
+    levelname: str
+    filename: str
+    lineno: int
+    funcName: str
+    processIndex: int
+
+    @classmethod
+    def decode(cls, message: str):
+        assert bool(message), "empty message"
+        levelname, filename, lineno, funcName, processIndex = message.split(":")
+        return cls(
+            levelname=levelname,
+            filename=filename,
+            lineno=int(lineno),
+            funcName=funcName,
+            processIndex=int(processIndex),
+        )
+
+    def encode(self) -> str:
+        return f"{self.levelname}:{self.filename}:{self.lineno}:{self.funcName}:{self.processIndex}"
+
+    def __str__(self) -> str:
+        return self.encode()
+
+    @classmethod
+    def get(cls, levelname: str):
+        processIndex = PartialState().process_index
+
+        # Gets the callsite of this function.
+        callsite_frame = inspect.currentframe().f_back
+        # logged filename is basename
+        filename = os.path.basename(callsite_frame.f_code.co_filename)
+        funcName = callsite_frame.f_code.co_name
+        lineno = callsite_frame.f_lineno
+
+        return cls(
+            levelname=levelname.upper(),
+            filename=filename,
+            lineno=lineno,
+            funcName=funcName,
+            processIndex=processIndex,
+        )
+
+
+def assert_equals(value, expected, prefix=""):
```

Comment:
We don't need this, pytest already generates readable error messages.

Reply:
I'm not quite an expert in testing, so thanks for your comments. I have re-written the tests into two simple tests: 1) check the logged message is correct in a normal setting; and 2) check the logged message is correct when the user overrides `stacklevel` in a custom logger wrapper. These tests are most relevant to this PR, and hopefully they can serve as a starting point for future tests.